fix: Ensure drop_search_index_on_message_id patch works for PostgreSQL

0xsirsaif opened the pull request:
This commit includes the following changes to fix the failing patch:

1. Added `get_column_index` method to the `PostgresDatabase` class.
2. Modified the patch code to drop the index correctly for PostgreSQL.

The `get_column_index` method checks if a column exists for a specific field, and the patch now uses this method to drop the index on "message_id" when PostgreSQL is the backend database.

0xsirsaif committed:
fix: Ensure drop_search_index_on_message_id patch works for PostgreSQL

This commit includes the following changes to fix the failing patch:

1. Added `get_column_index` method to the `PostgresDatabase` class.
2. Modified the patch code to drop the index correctly for PostgreSQL.

The `get_column_index` method checks if a column exists for a specific field, and the patch now uses this method to drop the index on "message_id" when PostgreSQL is the backend database.

0xsirsaif commented:
@shariquerik Please could you check this?

ankush commented on frappe/database/postgres/database.py:
```diff
@@ -416,6 +416,34 @@ def get_column_type(self, doctype, column):
+
+		# TODO: check specific order, now it just checks if its single column index
+		for index in indexes:
+			# in case of single column index, indexname is same as fieldname
+			is_single_column_index = index.name == fieldname
```
This seems incorrect. 

We add suffix here: https://github.com/frappe/frappe/blob/0d5f28d569a5cac4bef92648c565a773e7c70a73/frappe/database/database.py#L1138-L1142

The function shouldn't rely on name of index at all IMO. That was the problem with previous mariadb implementation too. 


https://github.com/frappe/frappe/pull/18262

0xsirsaif committed:
feat: Add Python function to retrieve PostgreSQL index by column

- A new function `get_column_index` to find the index of a given field in a specific position within the index definition.
- Support optional filtering for unique indexes.
- Utilize PostgreSQL regular expressions for more robust index matching.

0xsirsaif commented:
@ankush I have reimplemented `get_column_index` and supported the option to return the index in a specific order

0xsirsaif committed:
lint: double quotes

akhilnarang commented on frappe/database/postgres/database.py:
```diff
@@ -416,6 +416,33 @@ def get_column_type(self, doctype, column):
 		return self.sql("SELECT datname FROM pg_database", pluck=True)
 
+	def get_column_index(
+		self, table_name: str, fieldname: str, unique: bool = False, order: int = 1
```
MariaDB variant doesn't take in `order` and no value is being passed when this is being called, so maybe would make sense to hardcode instead of the additional parameter?
